File: ios/OpenClawConsole/OpenClawConsole/Models/ApprovalRequest.swift
// Models/ApprovalRequest.swift
// OpenClaw Work Console
// ApprovalRequest model with action types matching protocol.md

import Foundation
import SwiftUI

// MARK: - Action Type

enum ApprovalActionType: String, Codable, CaseIterable {
    case deploy
    case shellCommand = "shell_command"
    case configChange = "config_change"
    case keyRotation = "key_rotation"
    case tradeExecution = "trade_execution"
    case destructive
    case askRootCause = "ask_root_cause"
    case proposeFix = "propose_fix"
    case acknowledge
    case gitCommit = "git_commit"
    case gitMerge = "git_merge"
    case gitPush = "git_push"
    case agentSkillInstall = "agent_skill_install"
    case agentRollback = "agent_rollback"

    var displayName: String {
        switch self {
        case .deploy: return "Deploy"
        case .shellCommand: return "Shell Command"
        case .configChange: return "Config Change"
        case .keyRotation: return "Key Rotation"
        case .tradeExecution: return "Trade Execution"
        case .destructive: return "Destructive Action"
        case .askRootCause: return "Root Cause"
        case .proposeFix: return "Propose Fix"
        case .acknowledge: return "Acknowledge"
        case .gitCommit: return "Git Commit"
        case .gitMerge: return "Git Merge"
        case .gitPush: return "Git Push"
        case .agentSkillInstall: return "Install Skill"
        case .agentRollback: return "Agent Rollback"
        }
    }

    var systemImage: String {
        switch self {
        case .deploy: return "arrow.up.circle"
        case .shellCommand: return "terminal"
        case .configChange: return "slider.horizontal.3"
        case .keyRotation: return "key"
        case .tradeExecution: return "chart.line.uptrend.xyaxis"
        case .destructive: return "trash"
        case .askRootCause: return "questionmark.circle"
        case .proposeFix: return "wrench.and.screwdriver"
        case .acknowledge: return "checkmark.circle"
        case .gitCommit: return "arrow.triangle.branch"
        case .gitMerge: return "arrow.triangle.merge"
        case .gitPush: return "arrow.up.doc"
        case .agentSkillInstall: return "square.and.arrow.down"
        case .agentRollback: return "arrow.uturn.backward"
        }
    }
}

// MARK: - Risk Level

enum RiskLevel: String, Codable {
    case high
    case critical

    var displayName: String {
        switch self {
        case .high: return "High"
        case .critical: return "Critical"
        }
    }

    var color: Color {
        switch self {
        case .high: return .orange
        case .critical: return .red
        }
    }
}

// MARK: - Approval Context

struct GitOperation: Codable, Hashable {
    let operationType: String
    let branchFrom: String?
    let branchTo: String?
    let commitMessage: String?
    let fileChanges: [String]?
    let diffSummary: String?

    enum CodingKeys: String, CodingKey {
        case operationType = "operation_type"
        case branchFrom = "branch_from"
        case branchTo = "branch_to"
        case commitMessage = "commit_message"
        case fileChanges = "file_changes"
        case diffSummary = "diff_summary"
    }
}

struct ApprovalContext: Codable, Hashable {
    let service: String
    let environment: String
    let repository: String
    let riskLevel: RiskLevel
    let gitOperation: GitOperation?

    enum CodingKeys: String, CodingKey {
        case service
        case environment
        case repository
        case riskLevel = "risk_level"
        case gitOperation = "git_operation"
    }
}

// MARK: - Approval Request

struct ApprovalRequest: Codable, Identifiable, Hashable {
    let id: String
    let agentId: String
    let agentName: String
    let actionType: ApprovalActionType
    let title: String
    let description: String
    let command: String
    let context: ApprovalContext
    let createdAt: Date
    let expiresAt: Date

    enum CodingKeys: String, CodingKey {
        case id
        case agentId = "agent_id"
        case agentName = "agent_name"
        case actionType = "action_type"
        case title
        case description
        case command
        case context
        case createdAt = "created_at"
        case expiresAt = "expires_at"
    }

    var isExpired: Bool {
        expiresAt < Date()
    }

    var timeUntilExpiry: TimeInterval {
        expiresAt.timeIntervalSince(Date())
    }

    static func == (lhs: ApprovalRequest, rhs: ApprovalRequest) -> Bool {
        lhs.id == rhs.id
    }

    func hash(into hasher: inout Hasher) {
        hasher.combine(id)
    }
}

// MARK: - Approval Response (outbound)

struct ApprovalResponse: Codable {
    let approvalId: String
    let decision: ApprovalDecision
    let biometricVerified: Bool
    let respondedAt: Date

    enum CodingKeys: String, CodingKey {
        case approvalId = "approval_id"
        case decision
        case biometricVerified = "biometric_verified"
        case respondedAt = "responded_at"
    }
}

// MARK: - Approval Decision

enum ApprovalDecision: String, Codable {
    case approved
    case denied
}

// MARK: - Governance

enum AgentPlanStepStatus: String, Codable {
    case pending
    case running
    case done
    case blocked
    case skipped
}

struct AgentPlanStep: Codable, Identifiable, Hashable {
    let id: String
    let agentId: String
    let title: String
    let details: String
    let status: AgentPlanStepStatus
    let owner: String?
    let evidence: [ResourceLink]
    let createdAt: Date
    let updatedAt: Date

    enum CodingKeys: String, CodingKey {
        case id
        case agentId = "agent_id"
        case title
        case details
        case status
        case owner
        case evidence
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }
}

struct EnvironmentObservation: Codable, Identifiable, Hashable {
    let id: String
    let agentId: String
    let source: String
    let summary: String
    let observedAt: Date

    enum CodingKeys: String, CodingKey {
        case id
        case agentId = "agent_id"
        case source
        case summary
        case observedAt = "observed_at"
    }
}

struct RollbackPoint: Codable, Identifiable, Hashable {
    let id: String
    let agentId: String
    let actionType: ApprovalActionType
    let title: String
    let description: String
    let command: String
    let createdAt: Date

    enum CodingKeys: String, CodingKey {
        case id
        case agentId = "agent_id"
        case actionType = "action_type"
        case title
        case description
        case command
        case createdAt = "created_at"
    }
}

struct GovernanceEvent: Codable, Identifiable, Hashable {
    let id: String
    let agentId: String
    let type: String
    let title: String
    let summary: String
    let createdAt: Date
    let actor: String
    let riskLevel: RiskLevel?
    let approvalId: String?
    let taskId: String?
    let incidentId: String?
    let rollbackPointId: String?

    enum CodingKeys: String, CodingKey {
        case id
        case agentId = "agent_id"
        case type
        case title
        case summary
        case createdAt = "created_at"
        case actor
        case riskLevel = "risk_level"
        case approvalId = "approval_id"
        case taskId = "task_id"
        case incidentId = "incident_id"
        case rollbackPointId = "rollback_point_id"
    }
}

struct AgentGovernanceState: Codable, Hashable {
    let agentId: String
    let currentObjective: String?
    let objectiveUpdatedAt: Date?
    let plan: [AgentPlanStep]
    let environment: [EnvironmentObservation]
    let rollbackPoints: [RollbackPoint]
    let events: [GovernanceEvent]

    enum CodingKeys: String, CodingKey {
        case agentId = "agent_id"
        case currentObjective = "current_objective"
        case objectiveUpdatedAt = "objective_updated_at"
        case plan
        case environment
        case rollbackPoints = "rollback_points"
        case events
    }
}

File: android/app/src/main/java/com/openclaw/console/data/model/ApprovalRequest.kt
package com.openclaw.console.data.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonObject

@Serializable
enum class ActionType {
    @SerialName("deploy") DEPLOY,
    @SerialName("shell_command") SHELL_COMMAND,
    @SerialName("config_change") CONFIG_CHANGE,
    @SerialName("key_rotation") KEY_ROTATION,
    @SerialName("trade_execution") TRADE_EXECUTION,
    @SerialName("destructive") DESTRUCTIVE,
    @SerialName("ask_root_cause") ASK_ROOT_CAUSE,
    @SerialName("propose_fix") PROPOSE_FIX,
    @SerialName("acknowledge") ACKNOWLEDGE,
    @SerialName("git_commit") GIT_COMMIT,
    @SerialName("git_merge") GIT_MERGE,
    @SerialName("git_push") GIT_PUSH,
    @SerialName("agent_skill_install") AGENT_SKILL_INSTALL,
    @SerialName("agent_rollback") AGENT_ROLLBACK
}

@Serializable
enum class RiskLevel {
    @SerialName("high") HIGH,
    @SerialName("critical") CRITICAL
}

@Serializable
enum class ApprovalDecision {
    @SerialName("approved") APPROVED,
    @SerialName("denied") DENIED
}

@Serializable
data class ApprovalContext(
    val service: String = "",
    val environment: String = "",
    val repository: String = "",
    @SerialName("risk_level") val riskLevel: RiskLevel = RiskLevel.HIGH,
    @SerialName("git_operation") val gitOperation: GitOperation? = null
)

@Serializable
data class GitOperation(
    @SerialName("operation_type") val operationType: String,
    @SerialName("branch_from") val branchFrom: String? = null,
    @SerialName("branch_to") val branchTo: String? = null,
    @SerialName("commit_message") val commitMessage: String? = null,
    @SerialName("file_changes") val fileChanges: List<String>? = null,
    @SerialName("diff_summary") val diffSummary: String? = null
)

@Serializable
data class ApprovalRequest(
    val id: String,
    @SerialName("agent_id") val agentId: String,
    @SerialName("agent_name") val agentName: String,
    @SerialName("action_type") val actionType: ActionType,
    val title: String,
    val description: String,
    val command: String,
    val context: ApprovalContext,
    @SerialName("created_at") val createdAt: String,
    @SerialName("expires_at") val expiresAt: String
)

@Serializable
data class ApprovalResponse(
    @SerialName("approval_id") val approvalId: String,
    val decision: ApprovalDecision,
    @SerialName("biometric_verified") val biometricVerified: Boolean,
    @SerialName("responded_at") val respondedAt: String
)

@Serializable
enum class AgentPlanStepStatus {
    @SerialName("pending") PENDING,
    @SerialName("running") RUNNING,
    @SerialName("done") DONE,
    @SerialName("blocked") BLOCKED,
    @SerialName("skipped") SKIPPED
}

@Serializable
data class AgentPlanStep(
    val id: String,
    @SerialName("agent_id") val agentId: String,
    val title: String,
    val details: String,
    val status: AgentPlanStepStatus,
    val owner: String? = null,
    val evidence: List<ResourceLink> = emptyList(),
    @SerialName("created_at") val createdAt: String,
    @SerialName("updated_at") val updatedAt: String
)

@Serializable
data class EnvironmentObservation(
    val id: String,
    @SerialName("agent_id") val agentId: String,
    val source: String,
    val summary: String,
    @SerialName("observed_at") val observedAt: String,
    val metadata: JsonObject? = null
)

@Serializable
data class RollbackPoint(
    val id: String,
    @SerialName("agent_id") val agentId: String,
    @SerialName("action_type") val actionType: ActionType,
    val title: String,
    val description: String,
    val command: String,
    @SerialName("created_at") val createdAt: String,
    val metadata: JsonObject? = null
)

@Serializable
data class GovernanceEvent(
    val id: String,
    @SerialName("agent_id") val agentId: String,
    val type: String,
    val title: String,
    val summary: String,
    @SerialName("created_at") val createdAt: String,
    val actor: String,
    @SerialName("risk_level") val riskLevel: RiskLevel? = null,
    @SerialName("approval_id") val approvalId: String? = null,
    @SerialName("task_id") val taskId: String? = null,
    @SerialName("incident_id") val incidentId: String? = null,
    @SerialName("rollback_point_id") val rollbackPointId: String? = null,
    val metadata: JsonObject? = null
)

@Serializable
data class AgentGovernanceState(
    @SerialName("agent_id") val agentId: String,
    @SerialName("current_objective") val currentObjective: String? = null,
    @SerialName("objective_updated_at") val objectiveUpdatedAt: String? = null,
    val plan: List<AgentPlanStep> = emptyList(),
    val environment: List<EnvironmentObservation> = emptyList(),
    @SerialName("rollback_points") val rollbackPoints: List<RollbackPoint> = emptyList(),
    val events: List<GovernanceEvent> = emptyList()
)
